### PR TITLE
[spirv] Fix storage class for ptr-to-ptr in Vulkan 1.2.

### LINKED ISSUE #3340: [SPIR-V] StructuredBuffer and RWStructuredBuffer can't be used as function parameters in HLSL of ray tracing shaders
Here is the example:

[[vk::binding(0,0)]] StructuredBuffer<float4> gSBuffer;
[[vk::binding(1,0)]] RWStructuredBuffer<float4> gRWSBuffer;

void foo(StructuredBuffer<float4> pSBuffer)
{
}

[shader("raygeneration")]
void main()
{
    foo(gSBuffer);
}

The compilation of this shader is failed. But SPIRV-Cookbook.rst tells that StructuredBuffer and RWStructuredBuffer can be used as  
function parameters.

### PR DESCRIPTION
**Pull request by ehsannas**
Also fixes #3340

### CODE REVIEW

**Comment by max-traverse** [comment]
I was about to submit an issue about a crash during the compilation of a shader, but we found out that this PR solves our issue.
I created the following repro shader that breaks on master, but works on this branch
```
struct RenderResourceHandle {
    uint handle;

    uint index() { return this.handle & ((1 << 23) - 1); }
};

struct BindingsOffset {
    RenderResourceHandle bindingsOffset;
    uint userData0;
    uint userData1;
    uint userData2;
};

[[vk::push_constant]] BindingsOffset g_bindingsOffset;

struct Bindings {
    RenderResourceHandle _dummy0;             // 0
    RenderResourceHandle _dummy1;             // 1
    RenderResourceHandle _dummy2;             // 2
    RenderResourceHandle _dummy3;             // 3
    RenderResourceHandle flat_bucket_indices; // 4
    RenderResourceHandle _dummy4;             // 5
    RenderResourceHandle _dummy5;             // 6
    RenderResourceHandle _dummy6;             // 7
    RenderResourceHandle _dummy7;             // 8
    RenderResourceHandle _dummy8;             // 9
};
ByteAddressBuffer g_byteAddressBuffer[] : register(t0, space3);

#define byte_buffer(handle) g_byteAddressBuffer[handle.index()]

struct FlatBucketIndex {
    uint location;
    uint length;
};

[shader("intersection")] void main() {
    Bindings bnd = byte_buffer(g_bindingsOffset.bindingsOffset).Load<Bindings>(0);
    ByteAddressBuffer flat_bucket_indices = byte_buffer(bnd.flat_bucket_indices);
}
```

Specifically assigning to the `flat_bucket_indices` is what causes the crash during compilation. Sampling from the buffer directly without assigning it as a variable worked as a workaround for us.

**Comment by Jasper-Bekkers** [comment]
Further reduced version of the above code sample:

```
ByteAddressBuffer g_byteAddressBuffer[] : register(t0, space3);
[numthreads(1,1,1)]
void main() {
    ByteAddressBuffer flat_bucket_indices = g_byteAddressBuffer[0];
}
```

`dxc -spirv -Tcs_6_5 -fspv-target-env=vulkan1.2 -Emain -fvk-use-dx-layout code.hlsl`

**Comment by ehsannas** [comment]
Thanks @max-traverse @Jasper-Bekkers I've added this as an additional test case.

@jaebaek PTAL.

File: tools/clang/lib/SPIRV/RemoveBufferBlockVisitor.cpp [hunk 1]
```diff
@@ -71,15 +72,88 @@ bool RemoveBufferBlockVisitor::visitInstruction(SpirvInstruction *inst) {
...
+  return true;
+}
+
+bool RemoveBufferBlockVisitor::updateStorageClass(
+    const SpirvType *type, const SpirvType **newType,
+    spv::StorageClass *newStorageClass) {
+  if (!isa<SpirvPointerType>(type))
+    return false;
+
+  auto *ptrType = cast<SpirvPointerType>(type);
```
Comment (jaebaek) [inline tools/clang/lib/SPIRV/RemoveBufferBlockVisitor.cpp:92]: Can we change the above code
```
  if (!isa<SpirvPointerType>(type))
    return false;

  auto *ptrType = cast<SpirvPointerType>(type);
```
to
```
auto *ptrType = dyn_cast<SpirvPointerType>(type);
if (ptrType == nullptr)
  return false;
```
?

I think the performance might be almost similar, but `cast` looks unsafe for me. I mean your code is ok but later someone can update the code in a wrong way.

File: tools/clang/lib/SPIRV/RemoveBufferBlockVisitor.cpp [hunk 2]
```diff
@@ -71,15 +72,88 @@ bool RemoveBufferBlockVisitor::visitInstruction(SpirvInstruction *inst) {
...
+      ptrType->getStorageClass() != spv::StorageClass::StorageBuffer) {
+    *newType =
+        context.getPointerType(innerType, spv::StorageClass::StorageBuffer);
+    *newStorageClass = spv::StorageClass::StorageBuffer;
+    return true;
+  }
+
+  // For pointer-to-pointer cases (which need legalization), we could have a
+  // type like: _ptr_Function__ptr_Uniform_type_StructuredBuffer_float.
+  // In such cases, we need to update the storage class for the inner pointer.
```
Comment (jaebaek) [inline tools/clang/lib/SPIRV/RemoveBufferBlockVisitor.cpp:106]: This code handles pointer-to-pointer cases e.g., `_ptr_Function__ptr_Uniform_type_StructuredBuffer_float`.
What happens if it is deeper pointer-to-pointer-to-pointer-to... ?

Do we need a recursive function or a for-loop for deeper cases that we do not even statically know the depth?

File: tools/clang/lib/SPIRV/RemoveBufferBlockVisitor.cpp (hunk 1 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/RemoveBufferBlockVisitor.cpp]: I agree, this is a better way. Done.

File: tools/clang/lib/SPIRV/RemoveBufferBlockVisitor.cpp (hunk 2 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/RemoveBufferBlockVisitor.cpp:106]: In SPIR-V it's illegal to have pointer to pointers. We deliberately use the ptr-to-ptr scheme for cases that require spirv legalization (usually in the form of inlining). spirv-opt recognizes this pattern and performs legalization. So there should be no cases of more than 2 ptrs.

If for some reason there's a need for more than 2 levels in the future, then that would require a very large effort fixing many places in the codebase, including both dxc and spirv-opt.

**Review by jaebaek — APPROVED** [review]

**Pull request merged**